File: Sponsors/mw_streams_python/mws/sources/csv_source.py
from typing import TypeVar, Iterable

from csv import DictReader
from dataclasses import is_dataclass

from mws.operators.operator import Source


T = TypeVar("T")


class CSVSource(Source[T]):
    def __init__(self, row_class: type, filepath: str) -> None:
        super().__init__()

        if not is_dataclass(row_class):
            raise ValueError("")

        self.row_class = row_class
        self.filepath = filepath

    def _get_rows(self):
        with open(self.filepath, "r") as f:
            return [self.row_class(**row) for row in DictReader(f)]

    def start(self) -> None:
        rows = self._get_rows()
        for row in rows:
            self.emit(row)

        self.flush()

File: Sponsors/mw-streams-java/src/main/java/com/mwam/hackathon/sources/CSVSource.java
package com.mwam.hackathon.sources;

import com.opencsv.bean.CsvToBeanBuilder;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;

/**
 * Parse a CSV file and emit the rows as record.
 *
 * @param <T> Java class to parse rows to.
 */
public class CSVSource<T> extends Source<T> {
    final Class<T> rowClass;
    final String filename;

    public CSVSource(Class<T> rowClass, String filename) {
        super(new ArrayList<>());
        this.rowClass = rowClass;
        this.filename = filename;
    }

    private List<T> getRows() {
        try {
            return new CsvToBeanBuilder<T>(new FileReader(filename))
                .withType(rowClass)
                .build()
                .parse();
        } catch (FileNotFoundException e) {
            // not good practice but unchecked exceptions are noisy
            throw new RuntimeException(e);
        }
    }

    /**
     * Begin emitting values.
     */
    @Override
    public void start() {
        for (final var row : getRows()) {
            emit(row);
        }

        flush();
    }
}
